lapce-data/src/selection_range.rs

```diff
@@ -0,0 +1,185 @@
+use lapce_rpc::buffer::BufferId;
+use lsp_types::{Range, SelectionRange};
+
+/// Lsp [selectionRange](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#selectionRange)
+/// are used to do "smart" syntax selection. A buffer id, buffer revision and cursor position are
+/// stored along side [`lsp_type::SelectionRange`] data to ensure the current selection still apply
+/// to the current buffer.
+#[derive(Clone, Debug)]
+pub struct SyntaxSelectionRanges {
+    pub buffer_id: BufferId,
+    pub rev: u64,
+    pub last_known_selection: Option<(usize, usize)>,
+    pub ranges: SelectionRange,
+    pub current_selection: Option<usize>,
+}
+
+/// Helper to request either the next or previous [`SyntaxSelectionRanges`],
+/// see: [`crate::editor::LapceUiCommand::ApplySelectionRange`]
+#[derive(Clone, Copy, Debug)]
+pub enum SelectionRangeDirection {
+    Next,
+    Previous,
+}
+
+impl SyntaxSelectionRanges {
+    /// Ensure the editor state match this selection range, if not
+    /// a new SelectionRanges should be requested
+    pub fn match_request(
+        &self,
+        buffer_id: BufferId,
+        rev: u64,
+        current_selection: Option<(usize, usize)>,
+    ) -> bool {
+        if self.last_known_selection.is_some() {
+            buffer_id == self.buffer_id
+                && rev == self.rev
+                && current_selection == self.last_known_selection
+        } else {
+            buffer_id == self.buffer_id && rev == self.rev
+        }
```

Comment:
Would something like this be too cryptic? 😅

```suggestion
        if buffer_id == self.buffer_id
                && rev == self.rev
                && (!self.last_known_selection.is_some()  || current_selection == self.last_known_selection)
```

Reply:
Still the common parts could be extracted into a variable, but no big deal :)